Update doc for KEP 2876 (#42120)
kubernetes/website: content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions.md

`messageExpression` is a CEL expression, so the restrictions listed in [Resource use by validation functions](#resource-use-by-validation-functions) apply. If evaluation halts due to resource constraints 
during `messageExpression` execution, then no further validation rules will be executed.

> #### The reason and fieldPath fields
> 
> Besides the `message` and `messageExpression` fields, which defines the string reported for a validation rule failure,
> we have also added two more fields under `validation`:
> - `reason` field which allows user to specify a machine-readable validation failure reason when a request fails this validation rule.
> - `fieldPath` field which specify the field path returned when the validation fails.

Reviewer: We like [timeless](https://developers.google.com/style/timeless-documentation) documentation: please avoid “we have also added”.

```suggestion

Setting `messageExpression` is optional.

#### The `message` field {#field-message}

If you want to set a static message, you can supply `message` rather than `messageExpression`.
The value of `message` is used as an opaque error string if validation fails.

Setting `message` is optional.

#### The `reason` field {#field-reason}

You can add a machine-readable validation failure reason within a `validation`, to be returned
whenever a request fails this validation rule.

<!-- an example here would be nice, but is hard to add using GitHub suggestions -->

Setting `reason` is optional.

#### The `fieldPath` field {#field-field-path}

You can specify the path to the data that have failed to validate.
```
NB. Can we explain _why_ it's useful to specify a `fieldPath`?

Author: Thank you for the review! I have updated the doc to align with the suggested format.
For `fieldPath`, we wanna user to be able to specify the accurate path when err returned instead of always using the path which `self` scoped. It is very usefully especially when do cross fields validation and user wanna the returned err indicate the path the err occurred. For example, if user trying to say a field has to be a different value than another field, something like `self.foo.test.a != self.b`. The validation has to be scoped high enough to have access to both field `a` and `b` but we wanna the returned err path to be on `a`, in this case having a fieldPath `.foo.test.a` will be great.
Hope that would make sense to you. Thank you^^

Author: Edit: I have also added explain in the doc with the example. Thank you

For example:

```yaml
x-kubernetes-validations:
- rule: "self.x <= self.maxLimit"
  messageExpression: '"x exceeded max limit of " + string(self.maxLimit)'
  reason: "FieldValueInvalid"
  fieldPath: ".x"
```

> The HTTP status code returned to the caller will match the reason of the reason of the first failed validation rule.

Reviewer 2: ```suggestion
The HTTP status code returned to the caller will match the reason of the first failed validation rule.
```

The currently supported reasons are: "FieldValueInvalid", "FieldValueForbidden", "FieldValueRequired", "FieldValueDuplicate".
If not set, default to use "FieldValueInvalid".

> For `fieldPath`, It must be a relative JSON path scoped to the location of this x-kubernetes-validations extension in the schema and refer to an existing field.

Reviewer 2: ```suggestion
The `fieldPath` value must be a relative JSON path that is scoped to the location of this x-kubernetes-validations extension in the schema. Additionally, it should refer to an existing field within the schema.
```

> For example when validation checks if a specific attribute `foo` under a map `testMap`, the fieldPath could be set to `".testMap.foo"` or `.tetsMap['foo']'`.  

Reviewer: ```suggestion
For example when validation checks if a specific attribute `foo` under a map `testMap`, you could set
`fieldPath` to `".testMap.foo"` or `.testMap['foo']'`. 
```

> If the validation checks two lists must have unique attributes, the fieldPath could be set to either of the list: e.g. `.testList1` or `.testList2`.

Reviewer 2: ```suggestion
If the validation requires checking for unique attributes in two lists, the fieldPath can be set to either of the lists. For example, it can be set to `.testList1` or `.testList2`.
```

> It supports child operation to refer to an existing field currently. Refer to [JSONPath support in Kubernetes](https://kubernetes.io/docs/reference/kubectl/jsonpath/) for more info.

Reviewer: ```suggestion
It supports child operation to refer to an existing field currently. Refer to [JSONPath support in Kubernetes](/docs/reference/kubectl/jsonpath/) for more info.
```

> And it does not support numeric index of array.

Reviewer: ```suggestion
The `fieldPath` field does not support indexing arrays numerically.
```

> 

Reviewer: ```suggestion

Setting `fieldPath` is optional.

```
